### DISCUSSION TITLE
String.prototype.matchAll called with a non-global RegExp argument

### DISCUSSION THREAD

**Discussion by airone01** [eslint/eslint#17246] [Ideas] [answered]
[This](https://github.com/eslint/eslint/assets/21955960/1ba02960-a712-4e58-8124-45decbad656b) happens when you run a `.matchAll(/regex/)` without your regex having the `g` flag.
I couldn't find a way to make it crash the compiler so that my code doesn't crash at runtime, but there's probably a way to make that happen that in Eslint (is there? I'm interrested).
Point is, if there isn't it's probably a good idea to invest time in it, and potentially consider a rule.
Another thing worth pointing out is that [the global flag behaves weirdly](https://stackoverflow.com/q/1520800/11609642) and can cause unexpected behavior. No idea if there is a way to inform the users of this and if Eslint is even the right place to discuss it, but why not, I'm not making a PR lol.

Also, I love what you guys are doing. You saved me a loooot of time.

**Comment by mdjermanovic** [comment]
> [This](https://github.com/eslint/eslint/assets/21955960/1ba02960-a712-4e58-8124-45decbad656b) happens when you run a `.matchAll(/regex/)` without your regex having the `g` flag.
> I couldn't find a way to make it crash the compiler so that my code doesn't crashes at runtime, but there's probably a way to make that happen that in Eslint (is there? I'm interrested).

You can configure the [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule to catch this, here's the selector:

```
CallExpression[callee.property.name='matchAll'] > Literal[regex.flags=/^[^g]*$/].arguments:first-child
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgc2VsZWN0b3I6IFwiQ2FsbEV4cHJlc3Npb25bY2FsbGVlLnByb3BlcnR5Lm5hbWU9J21hdGNoQWxsJ10gPiBMaXRlcmFsW3JlZ2V4LmZsYWdzPS9eW15nXSokL10uYXJndW1lbnRzOmZpcnN0LWNoaWxkXCIsXG4gICAgbWVzc2FnZTogXCJTdHJpbmcucHJvdG90eXBlLm1hdGNoQWxsIGNhbGxlZCB3aXRoIGEgbm9uLWdsb2JhbCBSZWdFeHAgYXJndW1lbnRcIlxufV0gKi9cblxuZm9vLm1hdGNoQWxsKC9yZWdleC8pIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczIwMjEiOnRydWV9fX0=)

**Reply by airone01** [reply]
Thanks, I'll do just that, but isn't worth considering it as a rule ?

**Reply by mdjermanovic** [reply, marked as answer]
We're not considering new core rules that target prototype methods, as core rules don't have type info to know that `foo.matchAll` is `String.prototype.matchAll` so it wouldn't be possible to avoid false positives. You can always create a custom rule.
